OpenRA.Mods.Common/Effects/RevealShroudEffect.cs

```diff
@@ -0,0 +1,82 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Effects;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Effects
+{
+	public class RevealShroudEffect : IEffect
+	{
+		static readonly PPos[] NoCells = { };
+
+		readonly WPos pos;
+		readonly Player player;
+		readonly Shroud.SourceType sourceType;
+		readonly WDist revealRadius;
+		readonly Stance validStances;
+		readonly int duration;
+
+		int ticks;
+
+		public RevealShroudEffect(WPos pos, WDist radius, Shroud.SourceType type, Player forPlayer, Stance stances, int delay = 0, int duration = 50)
+		{
+			this.pos = pos;
+			player = forPlayer;
+			revealRadius = radius;
+			validStances = stances;
+			sourceType = type;
+			this.duration = duration;
+			ticks = -delay;
+		}
+
+		void AddCellsToPlayerShroud(Player p, PPos[] uv) { if (!validStances.HasStance(p.Stances[player])) return; p.Shroud.AddSource(this, sourceType, uv); }
+
+		void RemoveCellsFromPlayerShroud(Player p) { p.Shroud.RemoveSource(this); }
+
+		PPos[] ProjectedCells(World world)
+		{
+			var map = world.Map;
+			var range = revealRadius;
+			if (range == WDist.Zero)
+				return NoCells;
+
+			return Shroud.ProjectedCellsInRange(map, pos, range)
+				.ToArray();
+		}
+
+		public void Tick(World world)
+		{
+			if (ticks++ < 0)
+				return;
+
+			if (ticks == 0)
```

Comment:
No because if (++ticks < 0) increments before checking if its lower then 0, so if its -1 it increments, checks, gives false and ticks is equal to 0 in the next if.
if (ticks++ < 0) increments after the < 0 check so -1 makes it true and when ticks is 0 it becomes 1 before the next if check.